boards/common/msba2/Makefile.include

```diff
@@ -19,6 +19,8 @@ PORT_LINUX ?= /dev/ttyUSB0
 # This does not make a lot of sense, but it has the same value as the previous code
 PORT_DARWIN ?= /dev/tty.usbserial-ARM
 
+# The -tg option is specific to pyterm so we are forced to use it.
+RIOT_TERMINAL = pyterm
```

Comment:
Can't you also use simply expanded variables `:=` here and in all following cases below? There's no need for recursions as it appears to me.

Reply:
At the moment we are trying to do the opposite: remove instances of `:=` where it is not needed.

Reply:
Interesting .. as far as I know you usually remove instances of `:`, where it is not needed (: maintaining recursively expanded variables is complex and can lead to hard-to-debug problems. Imagine someone replaces `pyterm` with a function or a variable in the future, but keeps the `=` ..

What's the motivation for choosing recursive assignments over simple assignments?

Reply:
`:=` has some disadvantages:

- The value of variables defined with `:=` **depends on the order they are defined**.
- Once introduced, it is not that easy to tell if `:=` can be replaced by `=`
- `:=` will be evaluated even if not needed. This will force evaluation of all called functions, in particular shell calls. This causes unnecessary overhead at minimum, or a failure that should not occur at most (think what happens if a tool is defined with `:=` but you don't have the tool and you don't need it either)

I believe immediate assignments cause harder-to-debug problems because the order of definition starts to matter.

With respect to performance, if some variable is too expensive to evaluate more than once, there is a trick to memoize it (I think there is a PR by @cladmi ).

Finally, recursive variables fit better into the declarative style that good makefiles should follow.

Reply:
a quote from the docs (:

```
To avoid all the problems and inconveniences of recursively expanded variables,
there is another flavor: simply expanded variables. 
```

Kidding aside, I am not very involved in the build system development and I don't know what the desired end result will look like. I was merely stating that I had problems with recursive evaluations in earlier projects and most of the time they lead to performance drops. If you think that recursive evaluations fit better in the picture, then I trust your judgement.

Reply:
The reality is that we (and by that I mean @cladmi and me) should probably spend some time writing a proper guide to writing makefiles in RIOT.

That part of the docs is true, but it refers mainly to things like wildcards and shell calls with side-efects. Essentially, I/O. In _our_ case `:=` is most of the times workaround for this. The real solution is to more cleanly separate the functional and I/O parts of our makefiles.

Performance drop in RIOT builds is mainly due to:

* FORCE and PHONY targets (for example, we are relinking every time).
  - these are ugly hacks because people could not figure out how to do dependencies properly.
* Recursive make (or more precisely, the fact that the submakes run _always_)
* Checking tools all the time / evaluating tool-related variables, even when those tools are not used (`:=` is involved here).
  - Here is where a memoization function will help, but only after we refactor.

This is probably more info than you asked for, but I feel this should be stated somewhere.

Reply:
If you are doing recursive definitions you are not defining but doing evaluation.

Here I would say there is no need for evaluation so no `:=`.

> Interesting .. as far as I know you usually remove instances of =, where it is not needed (: maintaining recursively expanded variables is complex and can lead to hard-to-debug problems. Imagine someone replaces pyterm with a function or a variable in the future, but keeps the = ..

`usually` I would say no, its only in RIOT that I saw this pattern.
If you have `INCLUDES` that depends on `INCLUDES` you should introduce a new name and do `INCLUDES = $(INCLUDES_TOOLCHAIN) $(INCLUDES_C_LIBRARY)` and define `INCLUDES_TOOLCHAIN` later instead of prepending to the previous value.

If you replace `pyterm` with a function, it does nothing as you are not evaluating the name with `$(pyterm)`.

> I was merely stating that I had problems with recursive evaluations in earlier projects and most of the time they lead to performance drops.

If a `shell` call is evaluated 100 times it can indeed slow down the build, but then only this one should be optimized. Also, with correct file targets you do not rebuild the target if not needed so do not re-evaluate the variable for nothing. It may make the `clean` build slower but makes incremental build faster.
Also, by doing `:=` everywhere you are making `make clean` slow as it checks the warnings supported for your `gcc` version even if you do not use it.

Reply:
@cladmi @jcarrano This discussion was quite insightful for me, I think it would be great to have this documented somewhere, kind ok like the `advanced-build-system-tricks.md` but a `build-system-basics.md` page :)

Reply:
It is somehow "gnumake" basics, but how it should be applied to `RIOT` could indeed be good.
I never had any issues before as was not using recursive makefiles.

We still need to `export` things and so how to counter the issues with it with `memoized` somehow and `target-export-variables`.